#Tratativa dos dados



In [52]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [53]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

In [54]:
#Dataset simples
!wget https://raw.githubusercontent.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/main/DADOS_ABERTOS_MEDICAMENTOS.csv

#Dataset com tajas
!wget https://github.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/blob/main/medTarja.xls

--2022-07-24 21:41:22--  https://raw.githubusercontent.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/main/DADOS_ABERTOS_MEDICAMENTOS.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5903386 (5.6M) [text/plain]
Saving to: ‘DADOS_ABERTOS_MEDICAMENTOS.csv.1’

DADOS_ABERTOS_MEDIC 100%[===================>]   5.63M  --.-KB/s    in 0.08s   

2022-07-24 21:41:22 (71.7 MB/s) - ‘DADOS_ABERTOS_MEDICAMENTOS.csv.1’ saved [5903386/5903386]

--2022-07-24 21:41:22--  https://github.com/lazarobodevan/TP-CienciaDosDados-3861-3865-3871/blob/main/medTarja.xls
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘m

In [55]:
pip install xlrd==1.2.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [56]:

df = pd.read_excel("/content/medTarja.xls", engine='xlrd', index_col=None, header=45, skiprows = [0, 45])
df

SUBSTÂNCIA                CNPJ  \
0                      CEFALOTINA SODICA  04.301.884/0001-75   
1                      CEFAZOLINA SÓDICA  04.301.884/0001-75   
2                      CEFOTAXIMA SÓDICA  04.301.884/0001-75   
3           CLORIDRATO DE CIPROFLOXACINO  04.301.884/0001-75   
4                CEFALEXINA MONOIDRATADA  04.301.884/0001-75   
...                                  ...                 ...   
27192  ACETATO DE RETINOL;COLECALCIFEROL  92.265.552/0001-40   
27193            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27194            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27195            ACETATO DE DEXAMETASONA  92.265.552/0001-40   
27196            ACETATO DE DEXAMETASONA  92.265.552/0001-40   

                                             LABORATÓRIO     CÓDIGO GGREM  \
0       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100106155118   
1       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100204157115   
2       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100310151117   
3       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100410113111   
4       AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA  500100801112111   
...                                                  ...              ...   
27192  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525917060050504   
27193  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525921120068107   
27194  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525921120067907   
27195  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525922030071603   
27196  MULTILAB INDÚSTRIA E COMÉRCIO DE PRODUTOS FARM...  525922030071703   

            REGISTRO          EAN 1          EAN 2       EAN 3  \
0      1516700060050  7898361881450         -           -        
1      1516700030038  7898361881405         -           -        
2      1516700050063  7898361881412         -           -        
3      1516700070099  7898361881313         -           -        
4      1516700080035  7898361880019         -           -        
...              ...            ...            ...         ...   
27192  1181901840031  7896472518821         -           -        
27193  1181902220010  7896472503599  7895296449267      -        
27194  1181902240011  7896472502110  7896004782560      -        
27195  1181902240021  7896004782225         -           -        
27196  1181902240038  7896004782232         -           -        

                            PRODUTO  \
0                 CEFALOTINA SÓDICA   
1                 CEFAZOLINA SÓDICA   
2                 CEFOTAXIMA SÓDICA   
3      CLORIDRATO DE CIPROFLOXACINO   
4                        CEFALEXINA   
...                             ...   
27192                       METOSAD   
27193       ACETATO DE DEXAMETASONA   
27194                      CORTITOP   
27195                      CORTITOP   
27196                      CORTITOP   

                                            APRESENTAÇÃO  ... PMC 18% ALC  \
0                        1G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
1                       1 G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
2                       1 G PO SOL INJ CT 50 FA VD TRANS  ...         NaN   
3              500 MG COM REV CT 2 BL AL PLAS TRANS X 07  ...       46,26   
4           500 MG COM REV CT BL AL PVC/PVDC TRANS X 8    ...       20,97   
...                                                  ...  ...         ...   
27192  50000 + 10000 UI/ML SOL OR CT FR PLAS AMB GOT ...  ...       27,91   
27193                 1 MG/G CREM DERM CT BG AL X 10 G    ...       12,30   
27194                 1 MG/G CREM DERM CT BG AL X 10 G    ...       19,05   
27195                   1 MG/G CREM DERM CT BG AL X 20 G  ...       37,92   
27196                   1 MG/G CREM DERM CT BG AL X 30 G  ...       56,86   

      PMC 20% RESTRIÇÃO HOSPITALAR  CAP CONFAZ 87 ICMS 0% ANÁLISE RECURSAL  \
0         NaN                  Sim  Não       Não     Não              

Tratando os dados para uma analise exploratória.

In [57]:
df.drop('EAN 2', inplace=True, axis=1)
df.drop('EAN 3', inplace=True, axis=1)
df.drop('ANÁLISE RECURSAL', inplace=True, axis=1)


In [58]:
df['PMC 17%'] = df['PMC 17%'].str.replace(',', '.')
df['PMC 18%'] = df['PMC 18%'].str.replace(',', '.')
df['PMC 20%'] = df['PMC 20%'].str.replace(',', '.')
df['PF 0%'] = df['PF 0%'].str.replace(',', '.')
df['PF Sem Impostos'] = df['PF Sem Impostos'].str.replace(',', '.')

df['PMC 17%'] = df['PMC 17%'].astype(float)
df['PMC 18%'] = df['PMC 18%'].astype(float)
df['PMC 20%'] = df['PMC 20%'].astype(float)
df['PF 0%'] = df['PF 0%'].astype(float)
df['PF Sem Impostos'] = df['PF Sem Impostos'].astype(float)

In [59]:
def map_tarjas(tarja):
  if tarja in ['Tarja Vermelha (*)','Tarja Vermelha(*)','Tarja  Vermelha', 'Tarja Vermelha', 'Tarja Vermelha sob restrição']:
    return 'Tarja Vermelha'
  if tarja in ['Tarja Preta(*)','Tarja  Preta']:
    return 'Tarja Preta'
  if tarja in ['Tarja Venda Livre/Sem Tarja (*)', 'Tarja Venda Livre', 'Tarja -(*)', '- (*) ','Tarja Sem Tarja']:
    return 'Sem Tarja'

df['TARJA'] = df['TARJA'].apply(map_tarjas)

No dataset utilizado para esse trabalho, possui 3 tipos de Tarja:

*   Tarja Vermelha
*   Tarja Preta
*   Sem Tarja

Na função acima, para evitar dublicatas, o nome das três tarjas foram padronizados.




#Regras de associação

Para a regra de associação, a coluna ['PF Sem Imposto'] foi classificada da seguinte forma:



*  Para medicamentos com o preço de R\$ 0 a  99,99 reais, foi padronizado como **Barato**.
*   Para medicamentos com o preço de R$ 100,00 a 499,99 reais, foi padronizado como **Médio**.

*   Para medicamentos com o preço de R$ 500,00 a 999,99 reais, foi padronizado como **Caro**.

*  Para medicamentos com o preço a partir de R$ 1000,00 reais, foi padronizado como **Muito Caro**.

Após esses parâmetros, foi criado a coluna **PF DISC**.


In [60]:
import sys
bins = [0, 100, 500,1000,max(df['PF Sem Impostos'])]
names = ['Barato', 'Médio', 'Caro', 'Muito caro']
df['PF DISC'] = pd.cut(df['PF Sem Impostos'], bins, labels=names)
#datasetMed.info()
df[['PF DISC','PF Sem Impostos']]

PF DISC  PF Sem Impostos
0           Médio           262.28
1            Caro           517.93
2      Muito caro          1917.29
3          Barato            27.44
4          Barato            12.44
...           ...              ...
27192      Barato            16.55
27193      Barato             7.30
27194      Barato            11.30
27195      Barato            22.49
27196      Barato            33.73

[27197 rows x 2 columns]

##Selecionando atributos relevantes

Analisado as colunas do Dataset utilizado.

In [61]:
df.columns

Index(['SUBSTÂNCIA', 'CNPJ', 'LABORATÓRIO', 'CÓDIGO GGREM', 'REGISTRO',
       'EAN 1', 'PRODUTO', 'APRESENTAÇÃO', 'CLASSE TERAPÊUTICA',
       'TIPO DE PRODUTO (STATUS DO PRODUTO)', 'REGIME DE PREÇO',
       'PF Sem Impostos', 'PF 0%', 'PF 12%', 'PF 17%', 'PF 17% ALC',
       'PF 17,5%', 'PF 17,5% ALC', 'PF 18%', 'PF 18% ALC', 'PF 20%', 'PMC 0%',
       'PMC 12%', 'PMC 17%', 'PMC 17% ALC', 'PMC 17,5%', 'PMC 17,5% ALC',
       'PMC 18%', 'PMC 18% ALC', 'PMC 20%', 'RESTRIÇÃO HOSPITALAR', 'CAP',
       'CONFAZ 87', 'ICMS 0%',
       'LISTA DE CONCESSÃO DE CRÉDITO TRIBUTÁRIO (PIS/COFINS)',
       'COMERCIALIZAÇÃO 2020', 'TARJA', 'PF DISC'],
      dtype='object')

Para a regra de Associação, foi criado a variavel **df_rules** que irá armazenar as colunas com mais relações, sendo elas:

*   LABORATÓRIO
*   CLASSE TERAPÊUTICA
*   TARJA 
*   PF DISC

Utlizando essas colunas, foi utilizado a função **get_dummies()** que é usad para manipulação de dados. Ele converte dados categóricos em variáveis ​​fictícias ou indicadoras



In [62]:
df_rules = df[['LABORATÓRIO','CLASSE TERAPÊUTICA', 'TARJA','PF DISC']]
df_rules.head()

LABORATÓRIO  \
0  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA   
1  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA   
2  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA   
3  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA   
4  AUROBINDO PHARMA INDÚSTRIA FARMACÊUTICA LIMITADA   

                 CLASSE TERAPÊUTICA           TARJA     PF DISC  
0  J1D2 - CEFALOSPORINAS INJETÁVEIS  Tarja Vermelha       Médio  
1  J1D2 - CEFALOSPORINAS INJETÁVEIS  Tarja Vermelha        Caro  
2  J1D2 - CEFALOSPORINAS INJETÁVEIS  Tarja Vermelha  Muito caro  
3      J1G1 - FLUORQUINOLONAS ORAIS  Tarja Vermelha      Barato  
4       J1D1 - CEFALOSPORINAS ORAIS  Tarja Vermelha      Barato

In [63]:
df_transactions = pd.get_dummies(df_rules)

In [64]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27197 entries, 0 to 27196
Columns: 813 entries, LABORATÓRIO_1FARMA INDUSTRIA FARMACEUTICA LTDA to PF DISC_Muito caro
dtypes: uint8(813)
memory usage: 21.1 MB


##Extraindo itemsets frequentes

Ulizando a função apriori:

In [94]:
frequent_itemsets = apriori(df_transactions, min_support = 0.1, use_colnames=True)
frequent_itemsets

support                                itemsets
0  0.173254                       (TARJA_Sem Tarja)
1  0.806486                  (TARJA_Tarja Vermelha)
2  0.618892                        (PF DISC_Barato)
3  0.234621                         (PF DISC_Médio)
4  0.135566       (TARJA_Sem Tarja, PF DISC_Barato)
5  0.470309  (TARJA_Tarja Vermelha, PF DISC_Barato)
6  0.204471   (TARJA_Tarja Vermelha, PF DISC_Médio)

Foi utilizado como suporte mínimo 0.1.

##Extraindo regras de associação

Extraindo as regras de associação, temos:

In [100]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.1)
rules.sort_values(by="confidence", ascending=False).head()

antecedents             consequents  antecedent support  \
5         (PF DISC_Médio)  (TARJA_Tarja Vermelha)            0.234621   
0       (TARJA_Sem Tarja)        (PF DISC_Barato)            0.173254   
3        (PF DISC_Barato)  (TARJA_Tarja Vermelha)            0.618892   
2  (TARJA_Tarja Vermelha)        (PF DISC_Barato)            0.806486   
4  (TARJA_Tarja Vermelha)         (PF DISC_Médio)            0.806486   

   consequent support   support  confidence      lift  leverage  conviction  
5            0.806486  0.204471    0.871493  1.080606  0.015252    1.505869  
0            0.618892  0.135566    0.782470  1.264309  0.028341    1.751982  
3            0.806486  0.470309    0.759922  0.942263 -0.028818    0.806045  
2            0.618892  0.470309    0.583159  0.942263 -0.028818    0.914276  
4            0.234621  0.204471    0.253533  1.080606  0.015252    1.025335


(PF DISC_Médio) -> (TARJA_Tarja Vermelha):
0.87 de confiança dos remedios com preço medio, possuem a tarja de cor vermelha, porem o suporte entre eles é de aproximadamente 20% e lift de 1.08.   


lift - o lift foi de 1.08, ou seja, podemos dizer que pf medio esta fortemente associado a tarja vermelha

In [67]:
def calcula_metricas(att_antecedente, att_consequente):
  print("(",att_antecedente,") -> (",att_consequente,")")

  suporte = len(df_transactions[(df_transactions[att_antecedente] == 1) & (df_transactions[att_consequente] == 1) ])/ \
  len(df_transactions)
  print("Suporte: ",suporte)

  confianca = len(df_transactions[(df_transactions[att_antecedente] == 1) & (df_transactions[att_consequente] == 1) ])/ \
  len(df_transactions[df_transactions[att_antecedente] == 1])
  print("Confiança: ",confianca)

  lift = confianca / (len(df_transactions[df_transactions[att_consequente] == 1]) / len(df_transactions))

  print("Lif: ",lift)

  print("--------------------------------")

calcula_metricas("PF DISC_Médio","TARJA_Tarja Vermelha")
calcula_metricas("TARJA_Sem Tarja","PF DISC_Barato")
calcula_metricas("PF DISC_Barato","TARJA_Tarja Vermelha")
calcula_metricas("TARJA_Tarja Vermelha","PF DISC_Barato")
calcula_metricas("TARJA_Tarja Vermelha","PF DISC_Médio")

( PF DISC_Médio ) -> ( TARJA_Tarja Vermelha )
Suporte:  0.20447108136926867
Confiança:  0.8714934963171916
Lif:  1.0806058456888237
--------------------------------
( TARJA_Sem Tarja ) -> ( PF DISC_Barato )
Suporte:  0.13556642276721698
Confiança:  0.7824702886247877
Lif:  1.264308723843177
--------------------------------
( PF DISC_Barato ) -> ( TARJA_Tarja Vermelha )
Suporte:  0.47030922528220026
Confiança:  0.7599215779467681
Lif:  0.9422625674942213
--------------------------------
( TARJA_Tarja Vermelha ) -> ( PF DISC_Barato )
Suporte:  0.47030922528220026
Confiança:  0.5831585666089176
Lif:  0.9422625674942213
--------------------------------
( TARJA_Tarja Vermelha ) -> ( PF DISC_Médio )
Suporte:  0.20447108136926867
Confiança:  0.25353332725449074
Lif:  1.0806058456888237
--------------------------------


In [68]:
rules[rules['confidence'] == rules.confidence.max()]

antecedents             consequents  antecedent support  \
5  (PF DISC_Médio)  (TARJA_Tarja Vermelha)            0.234621   

   consequent support   support  confidence      lift  leverage  conviction  
5            0.806486  0.204471    0.871493  1.080606  0.015252    1.505869

In [69]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Médio'})))].sort_values(by="lift",ascending=False)

antecedents      consequents  antecedent support  \
4  (TARJA_Tarja Vermelha)  (PF DISC_Médio)            0.806486   

   consequent support   support  confidence      lift  leverage  conviction  
4            0.234621  0.204471    0.253533  1.080606  0.015252    1.025335

In [70]:
rules[rules['consequents'].apply(lambda x: x.issubset(frozenset({'PF DISC_Barato'})))].sort_values(by="lift",ascending=False)

antecedents       consequents  antecedent support  \
0       (TARJA_Sem Tarja)  (PF DISC_Barato)            0.173254   
2  (TARJA_Tarja Vermelha)  (PF DISC_Barato)            0.806486   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.618892  0.135566    0.782470  1.264309  0.028341    1.751982  
2            0.618892  0.470309    0.583159  0.942263 -0.028818    0.914276